src/pip/_internal/cache.py

```diff
@@ -60,11 +62,23 @@ def _get_cache_path_parts(self, link):
             key_parts.append("=".join([link.hash_name, link.hash]))
         key_url = "#".join(key_parts)
 
+        # Include interpreter name, major and minor version in cache key
+        # to cope with ill-behaved sdists that build a different wheel
+        # depending on the python version their setup.py is being run on,
+        # and don't encode the difference in compatibility tags.
+        # https://github.com/pypa/pip/issues/7296
+        key = "{}-{}.{} {}".format(
```

Comment:
Since we'll be invalidating the cache anyway, it would be a good time to refactor the wheel cache key to something easier to describe and reproduce, in case we ever want to make that part of the public API. Something like sha224 of a canonicalized JSON object with keys "url", "hash", "interpreter", "hash name". Not sure if it would be better as a followup or a refactoring prior to this PR.

Reply:
Ok, I'll do that when #7335 is merged. It's probably easier to do it in an additional commit in this PR.